File: src/NeuralDamage.Domain/ChatMember.cs
using NeuralDamage.Domain.Enums;

namespace NeuralDamage.Domain;

public class ChatMember : BaseEntity
{
    public required Guid ChatId { get; init; }
    public Guid? UserId { get; init; }
    public Guid? BotId { get; init; }
    public ChatMemberRole Role { get; set; } = ChatMemberRole.Member;
    public DateTime JoinedAt { get; init; } = DateTime.UtcNow;

    public Chat Chat { get; set; } = null!;
    public User? User { get; set; }
    public Bot? Bot { get; set; }
}

File: backend/app/models/chat_member.py
from datetime import datetime
from uuid import uuid4

from sqlalchemy import String, DateTime, ForeignKey, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class ChatMember(Base):
    __tablename__ = "chat_members"
    __table_args__ = (
        UniqueConstraint("chat_id", "user_id", name="uq_chat_user"),
        UniqueConstraint("chat_id", "bot_id", name="uq_chat_bot"),
    )

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    chat_id: Mapped[str] = mapped_column(String(36), ForeignKey("chats.id"), nullable=False)
    user_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("users.id"), nullable=True)
    bot_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("bots.id"), nullable=True)
    role: Mapped[str] = mapped_column(String(20), default="member")
    joined_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    chat: Mapped["Chat"] = relationship("Chat", back_populates="members")
    user: Mapped["User | None"] = relationship("User")
    bot: Mapped["Bot | None"] = relationship("Bot")
